src/Orleans.Core/CodeGeneration/TypeUtils.cs

```diff
@@ -548,7 +549,34 @@ public static IEnumerable<Type> GetTypes(Assembly assembly, Predicate<Type> wher
             return assembly.IsDynamic ? Enumerable.Empty<Type>() : GetDefinedTypes(assembly, logger).Select(t => t.AsType()).Where(type => !type.GetTypeInfo().IsNestedPrivate && whereFunc(type));
         }
 
-        public static IEnumerable<TypeInfo> GetDefinedTypes(Assembly assembly, Logger logger)
+        public static IEnumerable<TypeInfo> GetDefinedTypes(Assembly assembly, ILogger logger)
+        {
+            try
+            {
+                return assembly.DefinedTypes;
+            }
+            catch (Exception exception)
+            {
+                if (logger != null && logger.IsEnabled(LogLevel.Warning))
+                {
+                    var message =
+                        $"Exception loading types from assembly '{assembly.FullName}': {LogFormatter.PrintException(exception)}.";
+                    logger.Warn(ErrorCode.Loader_TypeLoadError_5, message, exception);
+                }
+
+                var typeLoadException = exception as ReflectionTypeLoadException;
+                if (typeLoadException != null)
+                {
+                    return typeLoadException.Types?.Where(type => type != null).Select(type => type.GetTypeInfo()) ??
+                           Enumerable.Empty<TypeInfo>();
+                }
+
+                return Enumerable.Empty<TypeInfo>();
+            }
+        }
+
+        //TODO: delete this one after runtime migrate off Logger
+        public static IEnumerable<TypeInfo> GetDefinedTypes(Assembly assembly, Logger logger = null)
```

Comment:
why have this duplicate code, even if temporarily, just for the logger? Convert from one to the other and that's it, but we are still doing changes in these methods for 2.0.0 final, and it's easy to miss this unnecessary duplication

Reply:
that might bring much more changes , and I don't know how much more, since I need to migrate classes using GetDefinedTypes too and so on and so forth. I think the better approach is migrate classes using GetDefinedTypes (Logger) to GetDefinedTypes(ILogger). Then I can just delete GetDefinedTypes(Logger) at the end. 

I have gone on the route of what you suggested before, and that just end up with a spider web spread from the utility method using Logger to so much more. I don't want to go into the route again.